### PR TITLE
Use Multilock for Attester Slashing Protection

### PR DESCRIPTION
**Pull request by rauljordan**
**What type of PR is this?**

> Bug fix

**What does this PR do? Why is it needed?**

Validators running a lot of keys are running into deadlines due to lock contention when we update validator slashing protections. Unfortunately, we use a single mutex, so thousands of validators will need to wait to acquire a lock before being able to complete their attestations. This PR changes the behavior to use a multilock namespaced by public key.

### CODE REVIEW

File: validator/client/runner.go [hunk 1]
```diff
@@ -33,7 +33,6 @@ type Validator interface {
 	ProposeBlock(ctx context.Context, slot uint64, pubKey [48]byte)
 	SubmitAggregateAndProof(ctx context.Context, slot uint64, pubKey [48]byte)
 	LogAttestationsSubmitted()
-	ResetAttesterProtectionData()
```
Comment (rauljordan) [inline validator/client/runner.go:36]: Unused

File: validator/client/validator.go [hunk 2]
```diff
@@ -516,27 +516,18 @@ func (v *validator) UpdateProtections(ctx context.Context, slot uint64) error {
 	if err != nil {
```
Comment (rauljordan) [inline validator/client/validator.go:516]: This is fine as this function is only called in the main thread and never concurrently

**Review by terencechain — DISMISSED** [review]

File: validator/client/validator.go (hunk 2 above)
Comment (shayzluf) [reply validator/client/validator.go:516]: @rauljordan it is being called concurrently go routine at the end keep updating the old data while update protections is being called

File: validator/client/validator.go [hunk 3]
```diff
@@ -516,27 +517,27 @@ func (v *validator) UpdateProtections(ctx context.Context, slot uint64) error {
...
 	}
-	v.attesterHistoryByPubKeyLock.Lock()
+	newLocks := make(map[[48]byte]*mputil.Lock, len(attestingPubKeys))
+	for _, pubKey := range attestingPubKeys {
+		lock := mputil.NewMultilock(string(pubKey[:]))
+		newLocks[pubKey] = lock
+	}
+	for _, lock := range v.pkLocks {
+		lock.Lock()
+		defer lock.Unlock()
```
Comment (shayzluf) [inline validator/client/validator.go:527]: @rauljordan wdyt?
not the best pattern but the other way is to store oldLocks and iterate through them

**Comment by rauljordan** [comment]
Closing in favor of #8086

**Pull request closed without merging**